### ISSUE TITLE
Unable to lower the MTU

### ISSUE THREAD

**Issue by cespare** [tailscale/tailscale#8219]
### What is the issue?

I mostly cannot connect to other devices over Tailscale while my internet connection is via a wifi hotspot on my phone.

I say mostly because some requests sometimes work (such as a very tiny HTTP request). But most requests time out and SSH doesn't work.

I'm 99% certain that the problem is that the MTU on the hotspot or the cell network is lower than what Tailscale is using. I've previously debugged this issue when using plain wireguard and I was able to work around it by setting a lower MTU in wireguard. In Wireguard that was trivial: its config file has a setting named `MTU`; I lowered it, restarted Wireguard, and everything worked.

In Tailscale, I've been debugging for an hour or so and I can't get it working.

The [troubleshooting guide](https://tailscale.com/kb/1023/troubleshooting/?q=mtu#unable-to-make-a-tcp-connection-between-two-nodes) says:

> In order to solve this, we can set the MTU at the LAN level to a lower value, or use MSS (Maxiumum Segment Size) clamping..

but it does not say how to do these things.

I found some various iptables rules online and tried them but none seemed to work.

I learned about the `TS_DEBUG_MTU` setting (mentioned [here](https://github.com/tailscale/tailscale/wiki/Tailscale-1.10)). I set that in my `/etc/default/tailscaled`. I ran `tailscale down` and then `tailscale up --accept routes` to reload the config. There was no apparent effect, even at very low MTU (800). But also, I don't see anything about MTU in the logs, so it's really hard to tell if it's having an effect. I even tried introducing syntax errors in that config and got no warnings, so it's really unclear that it's doing anything whatsoever.

BTW that file did exist before I edited it. It had these contents:

```
# Set the port to listen on for incoming VPN packets.
# Remote nodes will automatically be informed about the new port number,
# but you might want to configure this in order to set external firewall
# settings.
PORT="41641"

# Extra flags you might want to pass to tailscaled.
FLAGS=""
```

(Also, this is a bit off-topic, but Tailscale's "sticky flags with hidden state" method of config is already pretty surprising and unlike all the other software I use, but adding a "normal" /etc config file *on top* is extra confusing.)

Questions:

* Is `TS_DEBUG_MTU` the best way to lower the MTU?
* How do I know whether my config file is being used?
* After I change my config, what is the easiest way to get tailscale to load it?
* How can I see what MTU tailscale is using?

### Steps to reproduce

Use a Linux laptop connected via wifi to a Pixel 6 on Google Fi with its wifi hotspot turned on.

Try to SSH over Tailscale.

### OS

Linux

### OS version

Arch

### Tailscale version

1.40.0-dev20230426

### Bug report

BUG-47d092bd1845c5e437e2126d67643b900553a343778b654b65580694a0f70f37-20230527011430Z-3d928383bfd85305

**Comment by raggi** [comment]
Tailscale always uses an MTU of 1280 on the tailscale0 interface. `$TS_DEBUG_MTU` can be used for debugging _at the moment_, but the environment variable controls are not intended for long running configuration, or for use outside of development or debugging - just be aware that it could go away sometime.

The first thing we need to do here is to diagnose what the root cause of your problem is. In your logs from the BUG link you provided I see evidence of 1280 byte (inside) packets flowing, which means that your outer network MTU is large enough.

You could validate that your connection has sufficient MTU by using a large ping to one of our derp servers, such as `ping -s 1400 -M dont 192.73.252.65`. If this works (while not using an exit node) you have more than enough space in your link MTU.

>(Also, this is a bit off-topic, but Tailscale's "sticky flags with hidden state" method of config is already pretty surprising and unlike all the other software I use, but adding a "normal" /etc config file on top is extra confusing.)

As I noted above, these debug options are not intended for normal configuration. The file `/etc/default/tailscaled` is a relatively normal path for Debian and alike distributions to host environment files for systemd services. In order for this file to be reloaded you would need to do `systemctl restart tailscaled`. `systemctl cat tailscaled | grep /default` will output the line from the systemd unit file that loads this configuration.

Regular configuration of tailscale is performed through `tailscale up` and `tailscale set`, and does not require restarting the daemon through systemctl.

**Comment by cespare** [comment]
> `$TS_DEBUG_MTU` can be used for debugging _at the moment_, but the environment variable controls are not intended for long running configuration, or for use outside of development or debugging - just be aware that it could go away sometime.

Okay, but it sounds like the easiest way to make this change right now. And tailscale certainly needs some way to lower the MTU, so hopefully if this option goes away it's because a better, permanent config knob is added.

> The first thing we need to do here is to diagnose what the root cause of your problem is. In your logs from the BUG link you provided I see evidence of 1280 byte (inside) packets flowing, which means that your outer network MTU is large enough.

Unfortunately I am in a different location now than I was earlier today and I cannot reproduce the issues. I'm pretty sure, based on prior experience, that it's because I've changed cell networks. I'll keep trying to debug during my travels over the coming week.

> You could validate that your connection has sufficient MTU by using a large ping to one of our derp servers, such as ping -s 1400 -M dont 192.73.252.65. If this works (while not using an exit node) you have more than enough space in your link MTU.

FWIW this ping command hangs right now when I'm on my wifi hotspot and works when I'm on a home internet connection. But my network has exit nodes configured. How do I go about "not using an exit node"?

> In order for this file to be reloaded you would need to do `systemctl restart tailscaled`. `systemctl cat tailscaled | grep /default` will output the line from the systemd unit file that loads this configuration.

Thanks, that's helpful. I can indeed see that systemctl says that my file is being used by tailscaled. However, it would be nice to have some indication at all that tailscale is actually seeing these settings. I can see nothing in the logs about my MTU setting or any other setting. And if I add what seems like a syntax error to the file

    foo bar "

the service restarts just fine with nothing mentioned in the logs or in `systemctl status tailscaled`. It's just all quite opaque. Is there a verbose mode I can turn on or something?

**Comment by cespare** [comment]
@DentonGentry this is Linux, not macos.

**Comment by DentonGentry** [comment]
> And tailscale certainly needs some way to lower the MTU, so hopefully if this option goes away it's because a better, permanent config knob is added.

1280 bytes is the minimum IPv6 MTU. I don't expect to provide a way to tune the production MTU to less than 1280 bytes. It is unlikely that a permanent config MTU knob will be added, though there are no plans to remove TS_DEBUG_MTU.

**Comment by raggi** [comment]
> FWIW this ping command hangs right now when I'm on my wifi hotspot

This is very likely indicating the problem that we need to diagnose for a better experience.

> However, it would be nice to have some indication at all that tailscale is actually seeing these settings. 

The current version of tailscaled will log this information:

```
$ grep MTU /etc/default/tailscaled
TS_DEBUG_MTU=1280
$ sudo systemctl restart tailscaled
$ sudo journalctl -eu tailscaled | grep MTU | tail -n1
May 28 16:16:11 lin tailscaled[1915088]: envknob: TS_DEBUG_MTU="1280"
```

> But my network has exit nodes configured. How do I go about "not using an exit node"?

The easiest option would be to de-configure the exit node temporarily to test, with `tailscale set --exit-node=""`, however if you do not want to do that you can use explicit interface selection in ping, so if your internet interface is `eth0` you would do:

```
ping -I eth0 -s 1400 -M do 192.73.252.65
```

> if I add what seems like a syntax error to the file the service restarts just fine with nothing mentioned in the logs or in systemctl status tailscaled. It's just all quite opaque. Is there a verbose mode I can turn on or something?

This behavior comes from systemd and is documented here: https://www.freedesktop.org/software/systemd/man/systemd.exec.html#EnvironmentFile=

You can also verify with the kernel exactly what environment variables were passed to the process with:

```
sudo cat /proc/$(systemctl show --property MainPID tailscaled | cut -d= -f2)/environ | xargs -n1 -0 echo
```

**Comment by wizzard0** [comment]
@raggi fyi I also keep the TS_DEBUG_MTU at 1200 as it's the only way to make it work with wifi networks and ISPs that keep MTU under 1400 (not under my control unfortunately)

**Comment by AmitAronovitch** [comment]
some update which might be related (not sure it would solves the problem for the OP)

TLDR: disabling IPV6 for the probematic connection (via NetworkManager UI) seems to solve the issue for me

Details:

I also had issues with ssh under tailscale in a specific location (I am assuming this is due to something on the ISP side there).
Specifically:
(1) ssh starts normally, but gets stuck when running commands with long/fancy output.
(2) ping -s 1400 also failed
(3) When activating tailscale after connecting to a commercial VPN service (which sets itself as default route), problem seems to disappear (but I wish to avoid using this extra VPN layer).

Trying to apply the TS_DEBUG_MTU approach described here, I noticed that it now breaks MagicDNS:  `/etc/resolv.conf`  does not get overwritten by tailscale/NM  (but 100.x addresses do seem to work), and the following message is shown the output `tailscale status`:

```
# Health check:
#     - router: multiple errors:
	no such device
	adding address xxxxxxxxxxx/128 from tunnel interface: invalid argument
```

Trying to work around this issue by disabling ipv6 for the connection did not work. But  - when I removed the TS_DEBUG_MTU setting and restarted tailscale,  I noticed that the original issue is gone :-)

Fortunately I can afford sticking to ipv4 when I am there, but it is possible that the "user-mode ipv6" feature of tailscale might still work (did not try).

**Comment by lethedata** [comment]
I think it might be worth mentioning to check (and how to check) the WAN/Gateway Interface MTU size under "[Troubleshooting Docs Unable to make a TCP connection between two nodes ](https://tailscale.com/kb/1023/troubleshooting#unable-to-make-a-tcp-connection-between-two-nodes)" to verify it's current size  and if it supports (larger than?) tailscale's MTU size and adjust if possible. 

I tan into a similar situation due to the default gateway being a GIF interface that was defaulting to an MTU size of 1280 even tho it supported a larger size.

**Comment by aglarendil** [comment]
@DentonGentry faced this with MacOS as well and IPv4, hence the argumentation wrt IPv6 MTU seems not completely correct. It would be great to have an option to tune MTU for a user without hacking env variables.

**Comment by luizbizzio** [comment]
Unfortunately, Tailscale does not provide an option to adjust the MTU. I created a repository where you can do this, whether on Windows or Linux. It forces the MTU to 1500 or any value you want.

Check it out here: https://github.com/luizbizzio/tailscale-mtu

**Comment by DennisGaida** [comment]
Using a similar approach that @luizbizzio is using, but just running a script at Windows logon.

I have a problem with running Tailscale in a Windows Hyper-V machine and TLS handshakes. TLS handshakes don't work with a particular reverse proxy in my network and I figured out it is tailscales MTU value. Lowering it to ~1180 makes the TLS handshakes work again (yes, disabling IPv6 at the same time) (`netsh interface ipv4 set subinterface "Tailscale" mtu=1180 store=persistent`).

I would love to debug further and figure out where the TLS handshake fails (HTTP connections work to that host, only SSL fails) - but I could narrow it down to Tailscale an its default 1280 MTU value, so right now it just works™️and I would also appreciate an MTU setting on Windows (since `TS_DEBUG_MTU` only works on Linux).

**Comment by Keno** [comment]
Though I'd add some breadcrumbs here in case someone googles the symptoms and finds the issue. I am on MacOS and am currently on train wifi (České dráhy - Czech national rail company). The symptom I observed is that SSH is stuck SSH2_MSG_KEXINIT`. Apparently this network has an max MTU of 1200 and silently drops packet above that:
```
$ ping -D -s 1172 8.8.8.8
PING 8.8.8.8 (8.8.8.8): 1172 data bytes
1180 bytes from 8.8.8.8: icmp_seq=0 ttl=118 time=39.573 ms
1180 bytes from 8.8.8.8: icmp_seq=1 ttl=118 time=34.153 ms

$ ping -D -s 1173 8.8.8.8
PING 8.8.8.8 (8.8.8.8): 1173 data bytes
556 bytes from 10.200.0.1: frag needed and DF set (MTU 1200)
Vr HL TOS  Len   ID Flg  off TTL Pro  cks      Src      Dst
 4  5  00 b104 0000   0 0000  40  01 b29d 10.200.104.215  8.8.8.8
```
This appears to translate to a max ping size of 1108 (I guess MTU of 1136?)
```
$ ping -s 1108 [tailscale host]
PING [tailscale host] (100.64.0.85): 1108 data bytes
1116 bytes from 100.64.0.85: icmp_seq=0 ttl=64 time=171.911 ms
1116 bytes from 100.64.0.85: icmp_seq=1 ttl=64 time=165.094 ms

$ ping -s 1109 [tailscale host]
PING [tailscale host] (100.64.0.85): 1109 data bytes
Request timeout for icmp_seq 0
Request timeout for icmp_seq 1
Request timeout for icmp_seq 2
```

I then did `sudo ifconfig utun4 mtu 1136` and ssh started working. I also wanted to try PTMU (#311), but I'm on the App store version of tailscale so as far as I could tell, there's no great way to enable TS_DEBUG_ENABLE_PMTUD.

**Comment by lethedata** [comment]
> This appears to translate to a max ping size of 1108 (I guess MTU of 1126?)

@Keno Max size there should be 1136 = 1108 Packet Size + 28 (8 bytes for ICMP header + 20 bytes for ethernet header).

**Comment by Keno** [comment]
> [@Keno](https://github.com/Keno) Max size there should be 1136 = 1108 Packet Size + 28 (8 bytes for ICMP header + 20 bytes for ethernet header).

Thanks, yes typo. Fixed.
